### ISSUE TITLE
EnumerableCollectionView broken in WPF due to Enumerator changes in Dictionary

### ISSUE THREAD

**Issue by batzen** [dotnet/wpf#9157]
### Description

https://github.com/dotnet/runtime/pull/82499 and then the follow up changes that introduced `GenericEmptyEnumerator` have completely broken `EnumerableCollectionView` in WPF for the case of initially empty dictionaries.

The issue here is that `EnumerableCollectionView` obtains an enumerator during construction and caches it.
When asked to check for changes it invokes `MoveNext` on the cached enumerator and expects and exception to be thrown if the underlying list/collection/dictionary was modified.

When the dictionary has a count of 0 during construction the enumerator will never throw that exception as it's and instance of `GenericEmptyEnumerator`.

@stephentoub 

### Reproduction Steps

```
var test = new Dictionary<string, string>();

Assert.That(CollectionViewSource.GetDefaultView(test).Cast<object>().Count(), Is.EqualTo(0));

test.Add("TEST", "TEST");

Assert.That(CollectionViewSource.GetDefaultView(test).Cast<object>().Count(), Is.EqualTo(1));
```

### Expected behavior

Count obtained from the view should be 1 after adding an item

### Actual behavior

Count obtained from the view is 0 after adding an item

### Regression?

Yes. Worked prior to the changes from the mentioned PR.

### Configuration

.NET 8 and later

**Comment by miloush** [comment]
The documentation for [`MoveNext`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerator.movenext?view=net-8.0) clearly states that `InvalidOperationException` is thrown if the collection was modified after the enumerator was created, and that calls to `MoveNext` keep returning `false` after the last element until `Reset` is called.

At least these don't seem to be implementation detail or undefined behavior.

**Comment by miloush** [comment]
The remark does say though that `MoveNext` is undefined if the collection was changed (which seems to be a slight in contradiction of the above).

What do you suggest then @stephentoub instead? Checking for an instance of `GenericEmptyEnumerator`? That sounds more like an implementation detail than any of the above.

Either way, here is the relevant code:

https://github.com/dotnet/wpf/blob/22e5e01a1f413431d395eb9e5f125163208c0315/src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/Data/EnumerableCollectionView.cs#L570-L591

**Comment by stephentoub** [comment]
> The documentation for [`MoveNext`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerator.movenext?view=net-8.0) clearly states that `InvalidOperationException` is thrown if the collection was modified after the enumerator was created, and that calls to `MoveNext` keep returning `false` after the last element until `Reset` is called.

It states if you get such an exception, the reason may be because the collection was modified. It is in no way guaranteed to happen. Many enumerables don't do this, including arrays, concurrent collections, most custom collection types, collections empty at the time of enumeration, LINQ enumerations over empty collections, removals from Dictionary, and on and on.

> What do you suggest then @stephentoub instead

Not rely on exceptions from enumerators as a way to be notified of a collection changing. It has never been a valid indicator.

**Comment by miloush** [comment]
@stephentoub fair enough. "Not rely on something" doesn't provide an alternative, but I am guessing the alternative is already supported. This is basically a best effort to support collections that do not provide change notifications as a courtesy to the users and as already observed, does not work in all cases.

We either continue this best effort and special case this optimization (for example double zero count), or we give up on this completely and document it as unsupported. @batzen do you have a scenario where you rely on this behavior?

**Comment by batzen** [comment]
@miloush Just upgraded from .NET 6 to .NET 8 and some unit tests broke because the count on the view is never updated.
Not sure if the tests that broke represent real world uses, as those tests are for internal helper methods to get counts from `ICollectionView`.
I can't check our UI tests until about mid next year for complicated reasons (porting from VB6 to .NET, COM interop etc.).

Will try to find out if we really rely on `EnumerableCollectionView` next week.

Documenting it as unsupported is not enough, i guess.
It either has to work, or it must throw an exception when trying to create a collection view that would lead to `EnumerableCollectionView`.

Will do some tests with `EnumerableCollectionView` to find out what's working and what's not so we/the WPF team can make an informed decision on what should happen to that class.

**Comment by miloush** [comment]
If the other collection views just silently ignore changes as the comment suggests, it doesn't seem unreasonable to align to this behavior. 

I am afraid throwing for everything that does not support notifications would seriously limit the usefulness of `EnumerableCollectionView`. You wouldn't be able to set the items source to a LINQ query for example, or as in this case, dictionary. But looking forward to your analysis.

**Comment by huiliuss** [comment]
In my test, the View after executing the Refresh method can obtain the real Count (either in .Net6 or .Net8 environment).

```C#

var test = new Dictionary<string, string>();

 

            Assert.That(CollectionViewSource.GetDefaultView(test).Cast<object>().Count(), Is.EqualTo(0));

 

test.Add("TEST", "TEST");

 

var myView = CollectionViewSource.GetDefaultView(test);

 

myView.Refresh();

 

Assert.That(myView.Cast<object>().Count(), Is.EqualTo(1));

```

**Comment by batzen** [comment]
Judging from my tests my proposed solution is:
If `_pollForChanges` is true we completely rely on calls to `Refresh`.
I would also prefer to rename `_pollForChanges` to `_sourceIsEnumerableWithoutNotifications` (or something like that) as we will no longer passively poll for changes.

This means:
- remove `_trackingEnumerator` and all code that uses it (this also means removing `EnsureSnapshot` as the initial snapshot is already taken in the ctor)
- remove `TraceData.CollectionChangedWithoutNotification` (it already notes that support for non notifying collections is incomplete and will be removed in a future version)

Reasoning:
Using enumerable types that don't implement `INotifyCollectionChanged` can never work reliably (without a call to `Refresh`).
So instead of "faking" support for it we make calling `Refresh` mandantory.
Using, for example, a linq expression as the source for `EnumerableCollectionView` is already completely broken when not calling `Refresh`.
By making calls to `Refresh` mandantory we at least make the behavior consistent and no reliant on the implementation of the enumerator.
